Move the tracking of unsplittable tablets to metadata table

cshannon opened the pull request:
This adds a new column to store information for tracking unsplittable tablets in the metadata table instead of in memory. This information can be used by the tablet management iterator to know if a tablet needs to split by checking the column and comparing it to the current metadata and avoid unnecessarily trying to split a tablet that can't be split. The data stored includes a hash of the file set and the relevant config related to splits and if this changes then the iterator will try and split again.

The metadata is stored as Json and the configs included include split threshold, table max end row size, and max open files as well as the file set hash. If a split happens on the tablet in the future (because the metadata changed such a the file set and the iterator attempted to split again) the column will be cleaned up by fate in UpdateTablets repo.

For the schema, I wasn't sure of the best approach so I just went ahead and created a new Split column family and unspilttable column qualifier. I thought it might be useful in case we wanted to store more things in the future related to splits in the new column family, however I can easily change the schema if desired or store the new metadata as a column part of an existing column family or just make it a stand alone column family/qualifier like the UserCompactionRequested  or Merged column families.

I marked this as a draft for now as it still needs more tests (only one test was modified so far) to be added and some other clean up work plus as I said I wasn't sure if the schema was the best approach.

This closes #4177

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:
```diff
@@ -381,6 +384,11 @@ public boolean getHostingRequested() {
     return onDemandHostingRequested;
   }
 
+  public UnSplittableMetadata getUnSplittable() {
```
This comment is not related to this line.  Need to update the toString() method on this class.

keith-turner reviewed (commented):
Realized there may be cases where the unsplittable marker could just hang around forever, even when its no longer applicable. Something like the following

 * A tablet needs to split, but cannot so the new unsplittable column is written
 * A compaction happens that filters a lot of the tablets data.  The tablet is no longer a candidate for split because it shrunk.

In the case above the unsplittable marker does not get removed.  It does not cause any harm. The only problem with this case is if someone is actively scanning the metadata table looking for problematic tablets then they will find false postives.

keith-turner commented on server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:
```diff
@@ -69,7 +72,23 @@ public class TabletManagementIterator extends SkippingIterator {
+        .isPresent()) {
+      return false;
+    }
+
```
Maybe this method should always return true if unsplittable column is present and things changed.  This would give the split code in the manager a chance to remove the column if its not needed anymore.  Thinking of something like the following.

```java
if(unsplittable is present AND unsplittableFromTablet != computedUnsplittable){
   // when unsplittable column is out of date always return true, even if tablet is not over
   // split threshold.  This gives code in manager a chance to remove the column.
   return true;
} else if(unsplittable is absent AND tablet is over split threshold){
   return true;
}
```

cshannon replied:
Ah just saw this comment, this would take away the need from having the compaction code having to deal with the marker so I like this better than having the compaction code deal with cleaning things up as it handles all cases if something changes

cshannon commented:
> In the case above the unsplittable marker does not get removed. It does not cause any harm. The only problem with this case is if someone is actively scanning the metadata table looking for problematic tablets then they will find false postives.

I am thinking to solve this we could update the compaction code to remove the marker when tablets are finished compacting?

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:
```diff
@@ -69,7 +72,23 @@ public class TabletManagementIterator extends SkippingIterator {
+
+    // If current config/files match unsplittable metadata then we can't split
+    if (Optional
+        .ofNullable(tm.getUnSplittable()).filter(um -> um.equals(UnSplittableMetadata
```
I am switching this to just a normal null check anyways with some refactoring I am doing

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:
```diff
@@ -75,13 +88,17 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
-      // mgmt iterator can inspect this column and only try to split the tablet when something has
-      // changed.
+
+      // TODO: Do we care about conditional mutations here? I don't think it is required because
```
Latest update adds in a conditional mutation, i added the condition to require files be the same as i figure we should fail fast and retry vs having to submit another op later when it detects it's different, i added a comment about it

cshannon commented:
@keith-turner - This is ready for another review, I realized I had to add a recheck in the FindSplits repo to see if we still are above the split threshold if no split points are found and the marker already exists. This is because the tablet management iterator will now try and split anytime there is a metadata change for an unsplittable tablet and it's possible we no longer need to split (like after a compaction) and we need to clear the marker. Before this change the marker was only cleared after a split was done.

I also added tests for TabletManagement and constraints as well as a couple ITs. There is one IT that will test that the marker is set and then that it is cleared after splits run successfully and a second test that will test that the marker is set and then cleared if we no longer need to split such as the split threshold increasing but no split was done.

cshannon commented:
One thing to mention is now that FindSplits fate repo has to recheck if a split is needed to clear the marker, it requires having the same logic for checking the threshold in multiple spots so I am wondering if we should consolidate that. For example, now [SplitUtils](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitUtils.java#L215), [FindSplits](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java#L153), and [TabletManagementIterator](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java#L95) all do checks on the split threshold and it seems error prone if we change the logic in the future and one spot is updated and not the others.

Also, all of those spots now need to compute the size by iterating over the files so I was wondering if we should move this into `TabletMetadata` class itself so it could be cached and re-used. Currently I updated the SplitUtils [findSplits](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitUtils.java#L205)() method to now take the estimatedSize as an argument so it's not computed more than once in FindSplits but that seems a bit odd. I was wondering if maybe we should add a getter that can compute the size and cache it (we could use Suppliers.memoize()) and then if any calling code needs to get the size of the files it could just call something like `tabletMetadata.getFileSize()`. The supplier could default to just returning 0 if files are not loaded.

I was also trying to think if we could simplify having to recheck the threshold in the condition of no splits found and the marker exists. SplitUtils findSplits() is already called before that point and already does the threshold check (as noted above) but it doesn't return why it found no splits. Another option is to return a reason from that method for finding no splits (besides the empty set currently) as if the reason returned was the threshold wasn't met then we could avoid having to recheck.

The only other thing was I was trying to think of if there are more test cases we should write.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:
```diff
@@ -0,0 +1,142 @@
+
+  private static final Gson GSON = ByteArrayToBase64TypeAdapter.createBase64Gson();
+
+  private final long splitThreshold;
+  private final long maxEndRowSize;
+  private final int maxFilesToOpen;
+  private final HashCode filesHash;
```
Thinking it may simplify the code in this class if all we store is a hash.  Something like the following.  Would remove the need for the json code.
```suggestion
  // stores a base64 encoded murmur 128 bit hash splitThreshold,maxEndRowSize,maxFilesToOpen, set of files.
  private final String hashOfSplitParameters;
```

cshannon replied:
This would definitely simply things but can you think of a reason we may need to read the original values? (maybe logging?) This change is probably fine as I don't think we need to query those stored values individually again but figure I'd double check before changing it. We could also log the values that are used to create the hash (maybe as trace) so we could access it there.

keith-turner replied:
> can you think of a reason we may need to read the original values?

The only case I can think of currently is debugging a problem in the code and I think for that case your suggestion to log covers that.

> We could also log the values that are used to create the hash (maybe as trace) so we could access it there.

I think logging the inputs to the hash and the resulting hash would be very useful.  Then if there is a problem can search the logs for the hash.

cshannon replied:
Changed in in 6c3358b5783b19ec6d73c7f7a121d0b0493c4333

keith-turner reviewed (commented):
Still reviewing changes, posting comment made so far.

keith-turner commented:
>  I was wondering if maybe we should add a getter that can compute the size and cache it (we could use Suppliers.memoize()) and then if any calling code needs to get the size of the files it could just call something like tabletMetadata.getFileSize(). The supplier could default to just returning 0 if files are not loaded.

That sounds nice, could  throw an exception instead of return zero if files were not fetched.

keith-turner commented:
> One thing to mention is now that FindSplits fate repo has to recheck if a split is needed to clear the marker, it requires having the same logic for checking the threshold in multiple spots so I am wondering if we should consolidate that. For example, now [SplitUtils](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitUtils.java#L215), [FindSplits](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java#L153), and [TabletManagementIterator](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java#L95) all do checks on the split threshold and it seems error prone if we change the logic in the future and one spot is updated and not the others.

Consolidating that decision logic to a single class would be really nice and less error prone.

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:
```diff
@@ -66,22 +91,68 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
           tabletMetadata.getLogs().size());
```
This is preexisting problem, seems like it should return null here.

```suggestion
          tabletMetadata.getLogs().size());
          return null;
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java:
```diff
@@ -242,6 +257,138 @@ public void automaticSplitLater() throws Exception {
+      final int numOfMutations = 20;
+      try (BatchWriter batchWriter = client.createBatchWriter(tableName)) {
+        // Make the last place in the key different for every entry added to the table
+        for (int i = 0; i < numOfMutations; i++) {
+          data[data.length - 1] = (byte) i;
+          Mutation m = new Mutation(data);
+          m.put("cf", "cq", "value");
+          batchWriter.addMutation(m);
+        }
```
There is another condition that can cause a tablet to be unsplittable and that is if every key has the same row.  This test could exercise that condition while `testUnsplittableColumn()` excercises the other condition that causes an unsplittable tablet.  Attempted to modify the code to create this other condition below. This way we get a check that both situations cause the marker to appear.


```suggestion
        for (int i = 0; i < numOfMutations; i++) {
          Mutation m = new Mutation(data);
          m.put("cf", "cq"+i, "value");
          batchWriter.addMutation(m);
        }
```

cshannon replied:
@keith-turner - the latest commit efe30445ad1e69151bdac4e93ac57f05d05d8e4e updates this test by writing the same row (that is not too large to split) as suggested and writes enough data so we end up splitting the default tablet and then we end up with a tablet that is unsplittable due to this condition. The test then still bumps the threshold and verifies all the tablets get cleared of the marker and no more splits happen. See what you think..we could also keep the old test and add this as a 3rd test too.

keith-turner reviewed (commented):
Finished review everything.

cshannon commented:
@keith-turner - looks like your new test in #4323 caused this PR to break...which is good news as that was the intent. I will update this PR to add the column to the test

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   UpdateTabletsTest.checkColumns:100 The split code does not know how to handle UNSPLITTABLE ==> expected: <true> but was: <false>
[ERROR]   UpdateTabletsTest.testManyColumns:363 
  Unexpected method call TabletMetadata.getUnSplittable():
[INFO] 
[ERROR] Tests run: 56, Failures: 2, Errors: 0, Skipped: 0
```

cshannon commented on server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:
```diff
@@ -69,7 +71,25 @@ public class TabletManagementIterator extends SkippingIterator {
 
   private static boolean shouldReturnDueToSplit(final TabletMetadata tm,
-      final long splitThreshold) {
+      final Configuration tableConfig) {
```
The problem with trying to put it into SplitUtils is that class is part of the manager module and TabletManagementIterator is part of base module so it's not visible as base doesn't have a dependency on manager (it's reversed)

Maybe SplitUtils should be moved to base?

keith-turner replied:
> Maybe SplitUtils should be moved to base?

Moving it to server base SGTM

cshannon commented on core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java:
```diff
@@ -339,7 +344,47 @@ public void testCompactionRequestedColumn() {
+    SplitColumnFamily.UNSPLITTABLE_COLUMN.put(mutation, new Value(unsplittableMeta.toJson()));
+    TabletMetadata tm = TabletMetadata.convertRow(toRowMap(mutation).entrySet().iterator(),
+        EnumSet.of(UNSPLITTABLE), true, false);
+    assertEquals(unsplittableMeta, tm.getUnSplittable());
```
Test was updated in 6c3358b5783b19ec6d73c7f7a121d0b0493c4333

cshannon commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:
```diff
@@ -66,22 +91,68 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
+          var mutator = tabletsMutator.mutateTablet(extent).requireAbsentOperation()
+              .requireSame(tabletMetadata, FILES).setUnSplittable(unSplittableMeta);
+          mutator.submit(tm -> unSplittableMeta.equals(tm.getUnSplittable()));
+        } else {
+          // We no longer need to split so we can clear the marker.
+          log.info("Tablet {} no longer needs to split, deleting unsplittable marker.",
+              tabletMetadata.getExtent());
+          var mutator = tabletsMutator.mutateTablet(extent).requireAbsentOperation()
+              .requireSame(tabletMetadata, FILES).deleteUnSplittable();
+          mutator.submit(tm -> tm.getUnSplittable() == null);
+        }
```
Good catch, this avoids having try try and clear the marker if it is already null and never set

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:
```diff
@@ -316,6 +322,11 @@ public Map<StoredTabletFile,DataFileValue> getFilesMap() {
     return files;
   }
 
+  public long getFileSize() {
```
```suggestion
   /**
    * @return the sum of the tablets files sizes
    */
  public long getFileSize() {
```


Would be nice to add a unit test for this new method.

cshannon replied:
Yeah, I still need to do some more testing updates for the ITs as well so i'll add it at the same time

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  private static HashCode caclulateSplitParamsHash(long splitThreshold, long maxEndRowSize,
```
Not really sure if implementing this comment is worthwhile, but also can not think of any harm implementing it may cause.

If a tablet is merged and this column is present, then merge should remove it.  To make the hash itself more strict, wondering if we should add the extent into the hash calculation.  That way if a merge happened and the column was not removed, then the hash would still differ if the extent is the only thing that is not the same between the pre and post merge tablet.  Maybe this overkill though.


```suggestion
  private static HashCode caclulateSplitParamsHash(KeyExtent extent, long splitThreshold, long maxEndRowSize,
```

cshannon replied:
Seems fine to add the extent as well, there is no harm since it is supposed to be per tablet anyways.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:
```diff
@@ -0,0 +1,102 @@
+    return toUnSplittable(Base64.getDecoder().decode(base64HashOfSplitParameters));
+  }
+
+  public static UnSplittableMetadata toUnSplittable(byte[] hashOfSplitParameters) {
```
I was looking at the MetadataConstraint class and wondering why it did not call this method because that code has a byte array.  Looking into it I think this method does not handle base64.  This could be private because nothing external calls it and its different expecations for input is very subtle.  Could inline it instead of making it private.

```suggestion
  private static UnSplittableMetadata toUnSplittable(byte[] hashOfSplitParameters) {
```

keith-turner commented on core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:
```diff
@@ -0,0 +1,116 @@
+    var hasher = Hashing.murmur3_128().newHasher();
+    hasher.putBytes(serializeKeyExtent(keyExtent)).putLong(splitThreshold).putLong(maxEndRowSize)
+        .putInt(maxFilesToOpen);
+    files.stream().map(StoredTabletFile::getNormalizedPathStr).sorted()
```
Would the following also include the range associated with the file?  If so that may be good to include.

```suggestion
    files.stream().map(StoredTabletFile::getMetadata).sorted()
```

If this is changed, then could update the unit test to have some files with ranges.

cshannon replied:
This is a good catch, we definitely should include the full json and metadata so we have the path and any ranges

keith-turner commented on server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:
```diff
@@ -73,15 +95,59 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
+          // Case 3: The table config and/or set of files changed since the tablet mgmt iterator
+          // examined this tablet.
+        } else {
+          log.info("Tablet {} no longer needs to split, ignoring it.", tabletMetadata.getExtent());
```
I would lower the log level on this, unless there is a bug with splitting its not useful.


```suggestion
          log.debug("Tablet {} no longer needs to split, ignoring it.", tabletMetadata.getExtent());
```

keith-turner commented on server/manager/src/test/java/org/apache/accumulo/manager/tableOps/split/UpdateTabletsTest.java:
```diff
@@ -261,6 +261,7 @@ public void testManyColumns() throws Exception {
     EasyMock.expect(tabletMeta.getHostingRequested()).andReturn(true).atLeastOnce();
     EasyMock.expect(tabletMeta.getSuspend()).andReturn(suspendingTServer).atLeastOnce();
     EasyMock.expect(tabletMeta.getLast()).andReturn(lastLocation).atLeastOnce();
+    EasyMock.expect(tabletMeta.getUnSplittable()).andReturn(null).atLeastOnce();
```
Would be good to test the case where its present to ensure its deleted.

```suggestion
    UnSplittableMetadata usm = UnSplittableMetadata.toUnSplittable(origExtent, 1000, 1001, 1002, tabletFiles.keySet());
    EasyMock.expect(tabletMeta.getUnSplittable()).andReturn(usm).atLeastOnce();
```

Then later in the code where it checks what methods were called on the mutator could do the following.

```java
EasyMock.expect(tablet3Mutator.deleteUnSplittable()).andReturn(tablet3Mutator);
```

keith-turner commented on test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java:
```diff
@@ -242,6 +257,148 @@ public void automaticSplitLater() throws Exception {
+
+      // Make sure no splits occurred in the table
+      assertTrue(client.tableOperations().listSplits(tableName).isEmpty());
+
```
Would be nice to squeeze the following in here.

 * record the current unsplittable value
 * set Property.TABLE_MAX_END_ROW_SIZE to 101
 * Wait for the unsplittable value to change (assuming the tablet still needs to split but can not)
 * set Property.TABLE_MAX_END_ROW_SIZE to 500 and continue with rest of test

cshannon replied:
That's a good idea to test that, i will add that.

cshannon replied:
I added this to the test in 12ee73ebc5ecc244681fc49e5528045f330ba995
